cmd/opm/http/http.go

```diff
@@ -0,0 +1,454 @@
+	if data.hasher == nil {
+		data.hasher = fnv.New64a()
+	}
+	data.hasher.Write(m.Blob)
+	idx[m.Schema][m.Package] = data
+}
+
+func (idx schemaAndPackageIndex) Get(keys ...string) (*indexData, error) {
+	if len(keys) != 2 {
+		return nil, fmt.Errorf("schema and package index only supports two keys")
+	}
+	v := idx[keys[0]][keys[1]]
+	return &v, nil
+}
+
+type schemaAndNameIndex map[string]map[string]indexData
+
+func (idx schemaAndNameIndex) Update(m declcfg.Meta, begin int64, end int64) {
+	if _, ok := idx[m.Schema]; !ok {
+		idx[m.Schema] = map[string]indexData{}
+	}
+	data := idx[m.Schema][m.Name]
+	data.locations = append(data.locations, location{begin, end})
+	if data.hasher == nil {
+		data.hasher = fnv.New64a()
+	}
+	data.hasher.Write(m.Blob)
+	idx[m.Schema][m.Name] = data
+}
+
+func (idx schemaAndNameIndex) Get(keys ...string) (*indexData, error) {
+	if len(keys) != 2 {
+		return nil, fmt.Errorf("schema and name index only supports two keys")
+	}
+	v := idx[keys[0]][keys[1]]
+	return &v, nil
+}
+
+type packageAndNameIndex map[string]map[string]indexData
+
+func (idx packageAndNameIndex) Update(m declcfg.Meta, begin int64, end int64) {
+	if _, ok := idx[m.Package]; !ok {
+		idx[m.Package] = map[string]indexData{}
+	}
+	data := idx[m.Package][m.Name]
+	data.locations = append(data.locations, location{begin, end})
+	if data.hasher == nil {
+		data.hasher = fnv.New64a()
+	}
+	data.hasher.Write(m.Blob)
+	idx[m.Package][m.Name] = data
+}
+
+func (idx packageAndNameIndex) Get(keys ...string) (*indexData, error) {
+	if len(keys) != 2 {
+		return nil, fmt.Errorf("package and name index only supports two keys")
+	}
+	v := idx[keys[0]][keys[1]]
+	return &v, nil
+}
+
+type schemaPackageAndNameIndex map[string]map[string]map[string]indexData
+
+func (idx schemaPackageAndNameIndex) Update(m declcfg.Meta, begin int64, end int64) {
+	if _, ok := idx[m.Schema]; !ok {
+		idx[m.Schema] = map[string]map[string]indexData{}
+	}
+	if _, ok := idx[m.Schema][m.Package]; !ok {
+		idx[m.Schema][m.Package] = map[string]indexData{}
+	}
+	data := idx[m.Schema][m.Package][m.Name]
+	data.locations = append(data.locations, location{begin, end})
+	if data.hasher == nil {
+		data.hasher = fnv.New64a()
+	}
+	data.hasher.Write(m.Blob)
+	idx[m.Schema][m.Package][m.Name] = data
+}
+
+func (idx schemaPackageAndNameIndex) Get(keys ...string) (*indexData, error) {
+	if len(keys) != 3 {
+		return nil, fmt.Errorf("schema, package, and name index only supports three keys")
+	}
+	v := idx[keys[0]][keys[1]][keys[2]]
+	return &v, nil
+}
+
+func indexReader(r io.ReaderAt, idx index, keys ...string) (io.Reader, string, error) {
+	data, err := idx.Get(keys...)
+	if err != nil {
+		return nil, "", err
+	}
+	readers := make([]io.Reader, 0, len(data.locations))
+	for _, loc := range data.locations {
+		readers = append(readers, io.NewSectionReader(r, loc.begin, loc.end-loc.begin))
+	}
+	return io.MultiReader(readers...), fmt.Sprintf("%x", data.hasher.Sum(nil)), nil
```

Comment:
This is the coolest part of this PR, IMO.

Go stdlib has an `io.SectionReader` and an `io.MultiReader` that can be used in conjunction to produce a single reader that reads _just_ the locations of the file that our index tells us to read.

Reply:
Agree that this is very cool. I learned something today. Thanks!